### PR TITLE
fix(android): clean/rebuild should release gradle file locks

### PR DESCRIPTION
**Pull request by jquick-axway**
**JIRA:**
https://jira.appcelerator.org/browse/TIMOB-27882

**Summary:**
- Gradle daemon file locks prevented some build files from being deleted on Windows.
- Modified app rebuild to stop gradle daemon on Windows only.
- Modified module clean to stop gradle daemon on Windows only.

**Hyperloop App Test:**
1. Install this PR's SDK on a Windows 10 machine.
2. Download the [hyperloop-examples](https://github.com/appcelerator/hyperloop-examples) project.
3. Open "tiapp.xml" and remove version from hyperloop module entry.
`<module>hyperloop</module>`
4. Build with Titanium version `9.0.1.GA` for Android.
5. Build with this PR's SDK version for Android and verify it works.
6. Build with Titanium version `8.3.1.GA` for Android.
7. Build with this PR's SDK version for Android and verify it works.

**Module Clean Test:**
1. Install this PR's SDK on a Windows 10 machine.
2. Download the source for [ti.imagefactory](https://github.com/appcelerator-modules/ti.imagefactory).
3. Open the Command Prompt.
4. CD to folder: `.\ti.imagefactory\android`
5. Enter: `appc ti sdk select <sdk-version>` _(Enter this PR's SDK version.)_
5. Enter: `appc run -p android --build-only`
6. Enter: `appc ti clean`
7. Verify that the clean occurred without errors.
8. Verify folders were deleted: `.\android\build` and `.\android\dist`
9. Enter: `appc run -p android`
10. Verify module was built and example app runs in emulator.

### CODE REVIEW

File: android/cli/commands/_cleanModule.js [hunk 1]
```diff
@@ -12,14 +12,28 @@
...
-exports.run = function run(logger, config, cli, finished) {
+exports.run = async function run(logger, config, cli, finished) {
 	const projectDir = cli.argv['project-dir'];
 
+	// Run the gradle "clean" task if possible.
+	// This makes the gradle daemon release its file locks so that they can be deleted on Windows.
+	try {
+		const gradlew = new GradleWrapper(path.join(projectDir, 'build'));
+		gradlew.logger = logger;
+		if (await gradlew.hasWrapperFiles()) {
```
Comment (garymathews) [inline android/cli/commands/_cleanModule.js:30]: Upon testing, `hasWrapperFiles()` returns `false` and `appc ti clean` ends with
```
(node:21360) UnhandledPromiseRejectionWarning: Error: EBUSY: resource busy or locked, unlink 'C:\Users\gmathews\github\appcelerator.https\android\build\local.properties'
```

File: android/cli/commands/_cleanModule.js [hunk 2]
```diff
@@ -12,14 +12,28 @@
...
+			await gradlew.clean();
+		}
+	} catch (err) {
+		this.logger.error(`Failed to run gradle "clean" task. Reason:\n${err}`);
+	}
+
 	const toDelete = [ 'build', 'dist', 'java-sources.txt' ];
 	toDelete.forEach(f => {
```
Comment (garymathews) [inline android/cli/commands/_cleanModule.js:38]: We can make this async
```
async f => {
  ...
  await fs.remove(target);
```
Same for `arches.forEach(arch => {`

File: android/cli/commands/_cleanModule.js (hunk 1 above)
Comment (garymathews) [reply android/cli/commands/_cleanModule.js:30]: I'm not sure how my project got into that state, after manually killing the process that had the lock and re-testing, it's working fine now.

**Review by garymathews — APPROVED** [review]
CR: **PASS**

From my previous comments, I believe my test project was in an initial bad state. Subsequent builds are successful.

**Review by ssjsamir — COMMENTED** [review]
@jquick-axway The first test seems to be fine but on the `Module Clean Test` when I do a `appc ti clean` I see the following error:
```
[DEBUG]  Deleting  C:\Users\APPC\Downloads\ti.imagefactory-stable\android\build
(node:28180) UnhandledPromiseRejectionWarning: Error: EBUSY: resource busy or locked, unlink 'C:\Users\APPC\Downloads\ti.imagefactory-stable\android\build\local.properties'
    at Object.unlinkSync (fs.js:1052:3)
    at rimrafSync (C:\ProgramData\Titanium\mobilesdk\win32\9.1.0\node_modules\fs-extra\lib\remove\rimraf.js:254:15)
    at C:\ProgramData\Titanium\mobilesdk\win32\9.1.0\node_modules\fs-extra\lib\remove\rimraf.js:291:39
    at Array.forEach (<anonymous>)
    at rmkidsSync (C:\ProgramData\Titanium\mobilesdk\win32\9.1.0\node_modules\fs-extra\lib\remove\rimraf.js:291:26)
    at rmdirSync (C:\ProgramData\Titanium\mobilesdk\win32\9.1.0\node_modules\fs-extra\lib\remove\rimraf.js:281:7)
    at Object.rimrafSync [as removeSync] (C:\ProgramData\Titanium\mobilesdk\win32\9.1.0\node_modules\fs-extra\lib\remove\rimraf.js:252:7)
    at C:\ProgramData\Titanium\mobilesdk\win32\9.1.0\android\cli\commands\_cleanModule.js:42:7
    at Array.forEach (<anonymous>)
    at Object.run (C:\ProgramData\Titanium\mobilesdk\win32\9.1.0\android\cli\commands\_cleanModule.js:38:11)
(node:28180) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:28180) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

**Test Environment**
```
Operating System
  Name                        = Microsoft Windows 10 Pro
  Version                     = 10.0.18363
  Architecture                = 64bit
  # CPUs                      = 4
  Memory                      = 32.0GB

Node.js
  Node.js Version             = 12.16.1
  npm Version                 = 6.13.4

Appcelerator CLI
  Installer                   = 5.0.0
  Core Package                = 8.0.0

Titanium CLI
  CLI Version                 = 5.2.2
  node-appc Version           = 0.2.49

Java Development Kit
  Version                     = 10.0.1_10
```

**Comment by jquick-axway** [comment]
**Updated PR:**
- Removed gradle "clean" task from last commit.
- Now always stops gradle daemon on clean/rebuild, but only for Windows.
  * This appears to be the only full proof solution to remove file locks.
  * Now always logs this action. Need to since gradle might download libraries/tools which can take a long time.
- Reverted hyperloop back to 5.0.3.
  * We don't need to stop the gradle daemon on hyperloop side since app build will stop it 1st.
- Modified module clean task to perform all operations async.

**Comment by ssjsamir** [comment]
@jquick-axway The Hyperloop app is running fine but when cleaning ti.image factory I get the following error (although the build folder empties)

```
 [DEBUG]  Deleting  C:\Users\APPC\Downloads\ti.imagefactory-stable\android\build
 [ERROR]  :An error occurred during clean after 2s 421ms
 [ERROR]  :EPERM: operation not permitted, lstat 'C:\Users\APPC\Downloads\ti.imagefactory-stable\android\build\build_android.log'
```

**Comment by jquick-axway** [comment]
The `./build/build_android.log` file is generated by our CLI. Hmm...

**Comment by jquick-axway** [comment]
@ewanharris is going to take over this PR from here. He is going to look into deleting all files under the "build" directory except for the log file since the CLI is actively writing to it. Meaning that a "clean" will no longer completely delete the "build" folder anymore. We might also make a CLI "clean" change as well. Stay tuned!

**Review by ssjsamir — APPROVED** [review]
**FR Passed:** Tested using the two test cases mentioned in the Description above. 

**Test Environment**
```
Operating System
  Name                        = Microsoft Windows 10 Pro
  Version                     = 10.0.18363
  Architecture                = 64bit
  CPUs                        = 4
  Memory                      = 31.9GB
Node.js
  Node.js Version             = 12.16.1
  npm Version                 = 6.13.4
Appcelerator CLI
  Installer                   = 5.0.0
  Core Package                = 8.0.0
Titanium CLI
  CLI Version                 = 5.2.2
  node-appc Version           = 0.2.49
Java Development Kit
  Version                     = 10.0.1_10
  Java Home                   = C:\Program Files\Java\jdk-10.0.1
NDK Version : 20.1.5948944
```

**Pull request merged**